packages/react-aria-components/src/Collection.tsx

```diff
@@ -709,16 +709,41 @@ export function useCollectionChildren<T extends object>(props: CachedChildrenOpt
 }
 
 const ShallowRenderContext = createContext(false);
+const CollectionDocumentContext = createContext<Document<any, BaseCollection<any>> | null>(null);
 
-interface CollectionResult<C> {
-  portal: ReactNode,
-  collection: C
+export interface CollectionBuilderProps<C extends BaseCollection<object>> {
+  content: ReactNode,
+  children: (collection: C) => ReactNode,
+  createCollection?: () => C
 }
 
-export function useCollection<T extends object, C extends BaseCollection<T>>(props: CollectionProps<T>, initialCollection?: C): CollectionResult<C> {
-  let {collection, document} = useCollectionDocument<T, C>(initialCollection);
-  let portal = useCollectionPortal<T, C>(props, document);
-  return {portal, collection};
+export function CollectionBuilder<C extends BaseCollection<object>>(props: CollectionBuilderProps<C>) {
+  // If a document was provided above us, we're already in a hidden tree. Just render the content.
+  let doc = useContext(CollectionDocumentContext);
+  if (doc) {
+    return props.content;
+  }
+
+  // Otherwise, render a hidden copy of the children so that we can build the collection before constructing the state.
+  // This should always come before the real DOM content so we have built the collection by the time it renders during SSR.
+
+  // This is fine. CollectionDocumentContext never changes after mounting.
+  // eslint-disable-next-line react-hooks/rules-of-hooks
+  let {collection, document} = useCollectionDocument(props.createCollection);
+  return (
+    <>
+      <Hidden>
+        <CollectionDocumentContext.Provider value={document}>
```

Comment:
These are technically not required for collections that render collection children directly (like Menu), it's only needed for cases like Select and ComboBox that have intermediary elements. It shouldn't hurt anything but if there were performance issues with this we could theoretically optimize it.